src/librustdoc/html/static/js/search.js

```diff
@@ -256,231 +503,204 @@ window.initSearch = function(rawSearchIndex) {
                 return 0;
             });
 
-            for (i = 0, len = results.length; i < len; ++i) {
-                result = results[i];
+            for (i = 0, len = ar.length; i < len; ++i) {
+                result = ar[i];
 
                 // this validation does not make sense when searching by types
-                if (result.dontValidate) {
+                if (result.dontValidate || isType) {
                     continue;
                 }
                 var name = result.item.name.toLowerCase(),
                     path = result.item.path.toLowerCase(),
                     parent = result.item.parent;
 
-                if (!isType && !validateResult(name, path, split, parent)) {
+                if (!validateResult(name, path, nameSplit, parent)) {
                     result.id = -1;
                 }
             }
-            return transformResults(results);
-        }
 
-        function extractGenerics(val) {
-            val = val.toLowerCase();
-            if (val.indexOf("<") !== -1) {
-                var values = val.substring(val.indexOf("<") + 1, val.lastIndexOf(">"));
-                return {
-                    name: val.substring(0, val.indexOf("<")),
-                    generics: values.split(/\s*,\s*/),
-                };
-            }
-            return {
-                name: val,
-                generics: [],
-            };
+            return transformResults(ar);
         }
 
-        function checkGenerics(obj, val) {
+        // `defaultLev` is the value to return in case nothing happened.
+        function checkGenerics(obj, val, defaultLev) {
+            if (obj.length <= GENERICS_DATA || obj[GENERICS_DATA].length === 0) {
+                return val.generics.length === 0 ? defaultLev : MAX_LEV_DISTANCE + 1;
+            }
             // The names match, but we need to be sure that all generics kinda
             // match as well.
             var tmp_lev, elem_name;
-            if (val.generics.length > 0) {
-                if (obj.length > GENERICS_DATA &&
-                      obj[GENERICS_DATA].length >= val.generics.length) {
-                    var elems = Object.create(null);
-                    var elength = obj[GENERICS_DATA].length;
-                    for (var x = 0; x < elength; ++x) {
-                        if (!elems[obj[GENERICS_DATA][x][NAME]]) {
-                            elems[obj[GENERICS_DATA][x][NAME]] = 0;
-                        }
-                        elems[obj[GENERICS_DATA][x][NAME]] += 1;
+            if (val.generics.length > 0 && obj[GENERICS_DATA].length >= val.generics.length) {
+                var elems = {};
```

Comment:
```suggestion
                var elems = Object.create(null);
```

[`Object.create(null)` is not the same thing as the `{}` object literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create).

Alternatively, if you don’t want to use an object with no prototype, you can use an [ES6 Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map). Both of these are decent options; I just don’t want to be stuck debugging problems caused by magic members variables like `__proto__`.

Reply:
For now we are still at ES5. I think we'll upgrade to ES6 in a year or two (ES6 will be "old enough" at this point I think).

Reply:
What browsers do we support that don't have ES6?

Reply:
The support on mobile seems a bit "imperfect" based on https://caniuse.com/?search=es6.

Reply:
> The support on mobile seems a bit "imperfect" based on https://caniuse.com/?search=es6.

What do you mean? The only mobile browser I see that doesn't support ES6 is Opera Mini.

Reply:
No, you're absolutely right. I was looking at some details but we don't care about them. Sorry about that.

So basically, we just need to update our whole JS (and tests) to ES6.

Reply:
We can update it incrementally; we don't need to do it all at once ;)

Reply:
Yes, I'm planning to do it afterwards.

Reply:
We've now started using ES6 in various places, so we can come back to this recommendation and use `Object.create(null)`